From 8b3976de7913b634111366a21589496e83551455 Mon Sep 17 00:00:00 2001
From: Alex Williamson <alex.williamson@redhat.com>
Date: Fri, 13 Jun 2014 19:18:05 -0400
Subject: [iommu] amd: Fix interrupt remapping for aliased devices

Message-id: <20140613191744.13853.99115.stgit@bling.home>
Patchwork-id: 83947
O-Subject: [RHEL7.1/7.0.z kernel PATCH] iommu/amd: Fix interrupt remapping for aliased devices
Bugzilla: 1090139
RH-Acked-by: Myron Stowe <mstowe@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

Bugzilla: 1090139
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7578407
Upstream commit: e028a9e6b8a637af09ac4114083280df4a7045f1

An apparent cut and paste error prevents the correct flags from being
set on the alias device resulting in MSI on conventional PCI devices
failing to work.  This also produces error events from the IOMMU like:

AMD-Vi: Event logged [INVALID_DEVICE_REQUEST device=00:14.4 address=0x000000fdf8000000 flags=0x0a00]

Where 14.4 is a PCIe-to-PCI bridge with a device behind it trying to
use MSI interrupts.

Signed-off-by: Alex Williamson <alex.williamson@redhat.com>
Cc: stable@vger.kernel.org
Signed-off-by: Joerg Roedel <joro@8bytes.org>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/iommu/amd_iommu.c b/drivers/iommu/amd_iommu.c
index 72531f0..76e579a 100644
--- a/drivers/iommu/amd_iommu.c
+++ b/drivers/iommu/amd_iommu.c
@@ -3999,7 +3999,7 @@ static struct irq_remap_table *get_irq_table(u16 devid, bool ioapic)
  iommu_flush_dte(iommu, devid);
  if (devid != alias) {
   irq_lookup_table[alias] = table;
-  set_dte_irq_entry(devid, table);
+  set_dte_irq_entry(alias, table);
   iommu_flush_dte(iommu, alias);
  }
 
-- 
1.7.1